enhancements/insights/config-api.md

```diff
@@ -0,0 +1,296 @@
+
+## Design Details
+
+### On demand data gathering options
+
+The on demand data gathering can happen almost immediately when there is no active gathering in the progress, but when there is some in progress 
+then the question is what to do in this situation. There are following options:
+
+- schedule it - e.g using something like the `ScheduleGatheringAt` attribute where the user would provide some time in the future.
+  - Pros 
+    - no need to interrupt gathering in progress
+    - we could probably use the `RateLimiting` queue and add a new gathering to the queue with the `AddAfter` method. Otherwise we would need to define some waiting time (when there’s already some gathering in progress) and right now it’s difficult to estimate an average gathering time since the gatherers run in parallel
+  - Cons
+    - user has to define the scheduling time correctly in the future
+    - gathering doesn’t happen immediately
+- force it - e.g using a `ForceGatherinReason` attribute where the user would provide some string as reason (mentioned in the [Proposal](#proposal)). 
+  - pros
+    - [roviding a reason (e.g “Refresh Insights”) is simpler for configuration than providing a time
+    - gathering would be triggered almost immediately
+    - there is already similar precedent with `ForceRedeploymentReason` when redeploying static pods
+  - cons
+    - need to be able to interrupt gathering (if any in progress)
+
+### Open Questions
+
+- ~~Can the API define default values? How does it work or how is it implemented?~~ This is done by using `// +kubebuilder:default=`
+- Should the new time of forced gathering (if any defined by user) set a new time for the existing gathering interval? If I force gathering at time A, will the new automatic gathering be in time A + interval?
+
+### Test Plan
+
+The new config API will require updates of the existing Insights operator testsuite. The testsuite will be extended to cover the new API and respective attributes. There will be a new test checking the forced gathering feature. 
+
+### Graduation Criteria
+
+The plan is to introduce the first version of the new API as the `v1alpha1` version and once we gain some confidence and experience with the new API then
+we can move it to `v1beta1`.
+
+#### Dev Preview -> Tech Preview
+
+N/A
+
+#### Tech Preview -> GA
+
+N/A
+
+#### Removing a deprecated feature
+
+The existing configuration via `support` secret should be marked as deprecated. This should be mentioned in the OCP docs (see [Support procedures](#support-procedures))
+
+### Upgrade / Downgrade Strategy
+
+If a cluster uses some non-default Insights configuration (there is an existing `support` secret overriding some default attributes) during an upgrade,
+then the config will have to be set in the new API to provide the same functionality. This is a one-time action. When there is an existings `insights.config.openshift.io` kind in a cluster before upgrade, then there is no need to update the config.
```

Comment:
Do we remove that secret after migration or do other components use it? How would we avoid confusion as to which is authoritative if it needs to remain?

Reply:
The secret is used only by the Insights operator. I wouldn't remove it, because there's one edge case (I can add it here) and that is an option of basic authentication for the Ingress service. Most of the time the token from `pull-secret` is used, but there's this option so I think if we want to stay backward compatible then we should keep it. I think it's not good idea to define username/password in the config API...or is it acceptable? 

The authoritative point should therefore be the existence of the `insights.config.openshfit.io` kind. If it exists, it is an authoritative source of configuration, with one exception, when you want to use basic authentication for Ingress. I can explicitly mention this in the enhancement text.

Reply:
Yeah, credentials should say in the secret. Sounds good about insights.config.openshift.io being the authority otherwise.

Reply:
discussion elsewhere about this:
https://github.com/openshift/enhancements/pull/1037#issuecomment-1064272077

That secret as it exists today is effectively an api, and i think we need to continue to respect the values an admin puts in that secret, with them taking precedence over the new api that an admin may not know about/configure (for config fields that both the secret and this new api support, as @tremes notes).

there are a couple paths...copying the config from the secret to the insights api object is somewhat problematic though, whether or not you then remove the secret, or remove the config from the secret: 

if the config bits are migrated, then you've got dual-writers to the insights api object (the controller migrating the config bits, which needs to keep doing so if the admin comes back and updates the secret again, and the admin who may be editing the insights object directly), but there is a clear precedence that the config values coming from the secret "win"  (but it also means you need to keep reconciling the insights api object and stomping the values back to the ones coming from the secret, if the insights object changes).  

It's somewhat easier to reason about having the controller simply consume both the secret and the insights api object, but make the values from the secret take precedence where they conflict.  And if you take that approach, you don't have to do a migration at all. (but you have to support the secret as a config mechanism indefinitely.  Which, again, i think is true anyway if we really want to hold ourselves to strict api compatibility)

Reply:
I agree. Thanks for summary Ben.

Reply:
I updated the enhancement according to Ben's suggestions. In short we will keep the `support` secret as the way to configure the Insights and if there's a conflict (option defined in the secret as well as in the API) with the new config API then the config in the secret takes precedence before the config in the new API.